securedrop_client/gui/widgets.py

```diff
@@ -2810,8 +2838,12 @@ def update_conversation(self, collection: list) -> None:
                     self.add_file(conversation_item, index)
 
         # If any items remain in current_conversation, they are no longer in the
-        # source collection and should be removed.
+        # source collection and should be removed from both the layout and the conversation
+        # dict. Note that an item may be removed from the source collection if it is deleted
+        # by another user (a journalist using the Web UI is able to delete individual
+        # submissions).
         for item_widget in current_conversation.values():
+            self.current_messages.pop(item_widget.uuid)
```

Comment:
since we already have a source dict and a conversation wrapper dict, why do we need a current_messages dict?

Reply:
they each serve different purposes on different widgets: one is on the source list mapping individual sources to their `SourceWidget` in the source list, one is on the main view mapping individual sources to `SourceConversationWrapper` parent widgets, and this one on the conversation view which maps message/file/reply items to their corresponding widget. This dict serves the purpose of on a per-source basis tracking which widgets are displayed: when we draw the view, we only draw items that aren't already in the dict